### ISSUE TITLE
Rendering issue: icon-spin fa-spin shake/wobble (chrome firefox)

### ISSUE THREAD

**Issue by Katuva** [FortAwesome/Font-Awesome#671]
When using data-loading-text, for example:

data-loading-text="<i class='icon-refresh icon-spin'></i> Logging in..."

And then when the button is clicked and the "loading" property is set, the icon will wobble and shake as it spins.

Although, if you include the spinning icon on the button as normal, for example:

<button type="button" id="login_btn" class="btn btn-primary"><i class='icon-refresh icon-spin'></i> Log in</button>

It spins fine.

I've only encountered this bug on Firefox (currently version 18), works perfectly on Chrome.

**edit by tagliala**
Firefox related issue: #3451

**Comment by skoub** [comment]
The problem seems to come from the icon itself. If you take the Icomoon spinner version, it does'nt wobble at all. It works well on Chrome 39, Windows and IE11, Windows. I don't know for the other browsers.

http://jsfiddle.net/cxxzh0uf/

**Comment by tagliala** [comment]
@skoub 

The approach of icomoon of keeping icons all of the same size seems better for this purpose but it isn't enough.

http://jsfiddle.net/tagliala/cxxzh0uf/2/

![image](https://cloud.githubusercontent.com/assets/556268/5607297/091e0aa8-9456-11e4-966d-e356f6ac1cc4.png)

If I inspect the `:before` pseudoelement, I notice that fontawesome is "weird"

![image](https://cloud.githubusercontent.com/assets/556268/5607303/35639d30-9456-11e4-8ac5-93180d3cdce6.png)

![image](https://cloud.githubusercontent.com/assets/556268/5607305/46e3c864-9456-11e4-874e-7197284305b9.png)

I think that this is related to the font itself and there is nothing we can do via CSS.

Dave should take a look at the font and inspect fonts generated by Icomoon to try to understand what is the best way to fix this issue.

We can't accept pull requests with edits to the .otf file, but we will appreciate some help on this matter. If someone here is able to take a look at the font and identify the problem is very welcome

Sorry for my English, it is also late here :)

**Comment by luxzg** [comment]
@tagliala thanks for linking fa-clock-o issue here as well, it's the same issue as rest are mentioning.
I've noticed that fa-cog has the same problem, but it is not as noticeable in the same circumstances as fa-clock or some of the others mentioned above. i'll keep watching this thread and will report if I find any workarounds, but I'm pretty convinced it's the small imperfections in glyphs themselves that are surfacing only once rotating, otherwise they're too small to notice by naked eye

**Comment by TsuiJie** [comment]
I have met the same issue, is there a confirmed solution for this?

**Comment by tagliala** [comment]
@TsuiJie no, there isn't

**Comment by MichaelMaven** [comment]
I had a similar issue with the shaking/wobbling spin icons in IE10 and 11. 

My problem was that in IE 10/11 the scrollbars were showing and hiding (pulsing) as long as the spinner was visible.

what i found was that by setting the height and width of the icon tag to a set size (same size as the icon itself), the pulsing stopped. 

Hope it helps someone.
Cheers

**Comment by davegandy** [comment]
Okay, I've figured out the issue is in Font Squirrel conversion to webfont formats. Looks like it's munging some of the metrics. My plan:

- test with a fresh version from Font Squirrel
- if that doesn't work, roll my own conversion solution

**Comment by davegandy** [comment]
Okay, after a considerable amount of investigation, I put in a few fixes:
- Updated fa-spinner to have a better visual effect when spinning.
- Home-rolled a webfont conversion. Font Squirrel was definitely messing things up.
- Adjusted some font metrics that apparently have no effect on mac, but are heavily used by windows.
- Added animated examples to the visual test page (http://fontawesome.io/test/).

**The Results**
- Retina Mac: Flawless in Chrome, Safari, and Firefox
- Non-retina Mac: Much better in Chrome, Safari, and Firefox
- Windows 7: IE11 looks flawless, Chrome 39 still has issues, but I don't think it's any worse than it was

**One Caveat**
- At a few font sizes, there's a bit of a wobble in all browsers (Icomoon has the same issue). Basically, if you keep it to multiples and half multiples of the optimized resolution (e.g. 14px, 21px, 28px, 35px), the wobble goes away.

Since the results are much better than they were, I'm going to call this acceptable for 4.3, and keep the issue open for 5.0.

**Comment by davegandy** [comment]
One more update. I figured out what selected option was not working well for the Font Squirrel conversion. So I put those files back in. Leaving open.

**Comment by skoub** [comment]
@davegandy, i don't see any improvement on my side with Firefox & Chrome on windows. I didnt tried IE11. You can clearly see the wobble if you use steps in the animation. 

I understand that you cannot do much on the 4.x branch but is there any solid solution for the 5.x branch? 

```
.fa-spin2 {
	animation: fa-spin 1s infinite steps(8);
    -webkit-animation: fa-spin 1s infinite steps(8);
    -moz-animation: fa-spin 1s infinite steps(8);
    -o-animation: fa-spin 1s infinite steps(8);
}
```

**Comment by qJake** [comment]
For anyone else that made it this far, and is looking for some troubleshooting:

**Do not use `icon-spin` and `icon-fixed-width` together. This can cause the "wobble" effect.**

This turned out to be my problem (and my solution).

**Comment by tagliala** [comment]
@skoub 

IE11 is almost perfect, IMHO the best in windows.

The spinner icon is new and it should be better than before.

Take a look at the bottom of this **very heavy** page with all browsers: http://fontawesome.io/test/

Chrome and firefox are really bad at small sizes :disappointed:

**Comment by skoub** [comment]
with IE11, with fa-pulse, there's a little glitch. It's like if the spinner is moving 1px up and go down on the next animation step.

**Comment by bphamous** [comment]
It's working for me on IE 11 and Chrome 40.0.2214.115 m but it's wobbling in Firefox 35.0.1.

**Comment by BreadMaker** [comment]
Still noticeable in Firefox 36 for Linux.

The video is recorded at 50 FPS.

[![Font-awesome issue #671 firefox linux ](http://img.youtube.com/vi/b3D4T-ANH00/0.jpg)](http://www.youtube.com/watch?v=b3D4T-ANH00)

**Comment by natewiebe13** [comment]
Adding a `scaleZ(1.001)` fixes the issue for Firefox 36 on Linux.

**Comment by JakeCooper** [comment]
Not sure if my problem was the same as everyone's, it seemed to be more like issue #4017, but that got closed and redirected here, so I'll just post my solution here.

Seems if you simply add the following to the .fa-spin class.

    .fa-spin {
    text-align:center;
    }

Seemed to fix the wobbliness for me.

**Comment by isaksky** [comment]
Neither avoiding fa-fixed @qJake nor text-align:center @JakeCooper fixes it for me:
![filter](https://cloud.githubusercontent.com/assets/607986/6495688/92371b34-c289-11e4-906e-f8b78ce1a909.gif)

(Without fa-spin, there is no jitter).

**Comment by tagliala** [comment]
@isaksky could you please provide a fiddle showing this issue? I have an idea

**Comment by isaksky** [comment]
@tagliala Here you go:

http://codepen.io/isaksky/pen/JoBqoM

**Comment by tagliala** [comment]
Chrome 40.0.2214.115, Windows 7. Can't reproduce that issue.

Which browser is giving you that output?

**Comment by isaksky** [comment]
I'm not on my work computer right now, but I believe it is the latest normal Chrome on Windows 8. I can't normally reproduce it here on this mac in Chrome, but if I zoom in to 110% or 125% for example, it shows up. (There is also a problem where my my buttons stack vertically, but that isn't the problem I mean (I think I can fix that by using ems).). Does the jitter happen for you if you mess with the zoom?

**Comment by isaksky** [comment]
I just got it to happen at normal (100%) zoom on this mac too. Seems related to messing with zoom and/or switching browser tabs. Btw the issue also happens with ```$('#spin-pls').toggleClass("fa-spin") ```, (instead of toggling 'hide' on it when already spinning) which might make it a little more clear where the problem is.

**Comment by tagliala** [comment]
I will check tomorrow with my mac

**Comment by jesseangelo** [comment]
I have a workaround here for spinning the cog. Nice and smooth when you hover.

http://codepen.io/jesseangelo/pen/KwrqdM?editors=010

I made a keyframe animation that spins only 30 degrees (<code>rotateFa</code> below). Then I call that like this <code>animation: rotateFa .1s infinite;</code>

```
@keyframes rotateFa {
    from { transform: rotate(0deg); }
    to { transform: rotate(30deg); }
}
```
And the opposite to reverse it back
```
@keyframes unRotateFa {
  from { transform: rotate(30deg); }
  to { transform: rotate(0deg); }
}
```

**Comment by Haleddo** [comment]
Problem still exist.

**Comment by ypsilonmeister** [comment]
It still exist in the case of hardware acceleration disabled mode. Firefox 36.0.4.

**Comment by tagliala** [comment]
Please **do not** +1 or confirm that this issue still exist, because the issue is open and because we do not know how to fix.

Please take a look here: https://github.com/FortAwesome/Font-Awesome/issues/671#issuecomment-65381998

**Comment by Vusys** [comment]
Perhaps it would be a good idea to put a heads up on the [examples](http://fortawesome.github.io/Font-Awesome/examples/) page? The issue is big enough for me to be very wary of animated icons, but as per https://github.com/FortAwesome/Font-Awesome/issues/671#issuecomment-71056820, you wouldn't notice on a retina mac - even if you tested on all the major browsers.

**Comment by tagliala** [comment]
@Vusys I agree. You could send a PR against the master branch or tell me what to write and I will do it

**Comment by Vusys** [comment]
@tagliala done with pull request #6686 which adds the following warning below the examples:

>Some browsers on some platforms have issues with animated icons resulting in a jittery wobbling effect. See [issue #671](https://github.com/FortAwesome/Font-Awesome/issues/671) for examples and possible workarounds.

**Comment by tagliala** [comment]
@Vusys thanks!

**Comment by lukehutch** [comment]
I see a much worse wobble than just +/- 1 pixel due to antialiasing. There is some weird interaction between the spin transformation and other CSS transformations, so that sometimes the center of rotation is outside the spinner icon.

I haven't nailed down which CSS rule is causing this issue yet, but it looks like a case of erroneous transformation order. The order should be: translate center of icon to (0,0), then rotate by current angle, then translate (0,0) to screen position at which center icon is to be displayed.  Somewhere in there CSS is throwing another translation into the mix.

**Comment by tagliala** [comment]
> I haven't nailed down which CSS rule

I'm pretty confident that it is not a css rule and the issue is not related to FA itself

Take a look here: http://jsfiddle.net/tagliala/g0ngLhyr/

**Comment by skoub** [comment]
@tagliala, if it's not related to a css rule or FA, how do you explain that Icomoon don't have the wobble effect?

demo: http://jsfiddle.net/cxxzh0uf/

**Comment by skoub** [comment]
another exemple is the one made by google: http://www.getmdl.io/components/index.html#loading-section/spinner

**Comment by tagliala** [comment]
@skoub your fiddle updated to 4.3.0: http://jsfiddle.net/tagliala/cxxzh0uf/9/

> Take a look here: http://jsfiddle.net/tagliala/g0ngLhyr/

Note that this fiddle doesn't have webfonts at all and still shows the wobble

**Comment by tagliala** [comment]
@skoub google's implementation is not based on webfonts and IMHO that is the proper way to deal with this feature.

**Comment by andrevanberlo** [comment]
I'd like to share my experience here. 

Single icons:
I only have a problem with wobble with stacked icons. Single icons spin or pulse just fine

Stacked icons:
Apparently on my site the code below does not wobble in a title where the font is a bit larger, but does wobble when the font is smaller. The fa-pulse wobbles in both cases

<span class="fa-stack"><i class="fa fa-video-camera fa-stack-2x" style="color:orange"></i><i class="fa  fa-gear fa-spin fa-stack-1x fa-inverse" style="color:white; margin-left:-15%" ></i></span>

disclaimer: I am very new to all these things but as my site produces both wobbly and good results I thought you all might want to know.

**Comment by toastal** [comment]
@isaksky That wobble/blur with spinner has to do with hardware acceleration

Items that are being shown/hidden with the `.hide` class can have the browser hinted with the that content will be hidden

```css
#spin-pls {
  will-change: contents;
}
```

However, Gecko-based browser, like Firefox, are, at least to me, still having issues with jumping text in the dialog box.

```css
.job-order-dialog {
  will-change: transform;
}
```

works, but negatively effects Chrome by overcompensating and blurring the text all of the time. I came up with two possible solutions, the first just targets Firefox, the latter is a different acceleration, `filter: blur(0);` noted in this bug but is a bit hacky in my opinion--similar to `transform: translateZ(0);`--and could lead to negative outcomes in the future.

```css
@-moz-document url-prefix() {
  .job-order-dialog {
    will-change: transform;
  }
}
```

```css
.job-order-dialog {
  filter: blur(0);
}
```

**Comment by isaksky** [comment]
That fixes the issue, thanks @toastal ! Could this be worked into a general solution @tagliala ?

**Comment by tagliala** [comment]
@toastal do you mind to provide a jsfiddle showing your solution and comparing with basic fontawesome spin?

**Comment by mautematico** [comment]
I'm trying al examples you guys posted, and I can't reproduce in Firefox Aurora:
>20150722004007
>Mozilla/5.0 (X11; Linux x86_64; rv:41.0) Gecko/20100101 Firefox/41.0

I'm on debian Jessie, and because of some nvidia driver issues, I think I don't have hardware acceleration.

>GLXtest process failed (exited with status 1): GLX extension missing 
>Procesador WebGL	Bloqueado para tu tarjeta gráfica debido a problemas no resueltos del controlador.

Also, can't reproduce here:
http://shb.github.io/bootstrap-loading-screen/#icon:spinner

is it solved, or it's because I don't have hardware acceleration enabled?

**Comment by torcado194** [comment]
I have found another potential fix for this issue (running firefox 39) thanks to this post: http://www.useragentman.com/blog/2014/05/04/fixing-typography-inside-of-2-d-css-transforms/

I have edited the <a href="http://jsfiddle.net/tagliala/g0ngLhyr/">JSFiddle above</a> with the fix: http://jsfiddle.net/5z9om45L/

just put "perspective(1px)" in the transform property. I have not tested this extensively so I do not know what specifically this does or if it causes any other issues, but so far it looks fine. This has been one of my biggest issues with Firefox, so I thought I'd mention it here.

**Comment by moobee** [comment]
> I have edited the JSFiddle above with the fix: http://jsfiddle.net/5z9om45L/

It fixes the issue just fine for me.

**Comment by spiderr** [comment]
Here is an updated Fiddle that shows the wobbling on the compass. The second locate button shows a simple fa-spin with pronounced wobbling.

http://jsfiddle.net/Lpur9ek1/1/

**Comment by tagliala** [comment]
@spiderr apparently, `fa-lg` increases the wobble effect because it deals with line height and font size. Try to use a font size of 14px or multiples

**Comment by spiderr** [comment]
@tagliala thanks, great tip! fa-2x definitely better. There are also some odd artifacts (dirty pixels) around the circle. Here is a fiddle at fa-5X that shows them more clearly on the large blue button. This is a minute detail, but if someone ever cleans up the compass, that would be great. My hope is that when the compass spins, you could not tell the the circle was moving, only the inner compass needle. If I sit 2 meters from the screen with fa-5x, it looks great :-)

http://jsfiddle.net/mhqLvw79/1/

**Comment by jonathanparrish** [comment]
Not sure if this is the right thread, but..
In firefox, my "fa fa-cog fa-spin" stops spinning.  I use it in an ajax call with a modal.  It works fine in Safari and Chrome but firefox stops for some reason.  Any thoughts??

**app/javascripts/my_file.js**
$(document).ready(function() {
  $('#search_btn').on('click', function() {
    $('#results').empty();
    var myForm = $('#user_search');
    $("#spinner_modal").modal("show");
    $.ajax({
      url: myForm.attr('action'),
      type: "POST",
      data: myForm.serialize(),
      dataType: 'text',
      success: function(html){
        $("#spinner_modal").modal("hide");
      	$('#results').append(html);
      },
      error: function(xhr, satus, err) {
        $("#spinner_modal").modal("hide");
      },
      timeout: 60000
    });
    return false;
  });

**_my_partial.html.erb**
<div id="spinner_modal" class="modal fade bs-example-modal-sm text-center" tabindex="-1" role="dialog" aria-labelledby="mySmallModalLabel">
  <div class="modal-dialog modal-sm fa-4x">
    <div class="modal-content">
      <i class="fa fa-cog fa-spin"></i>
    </div>
  </div>
</div>

**Comment by tagliala** [comment]
@jonathanparrish please start a new issue. Do not worry to mention me, I will take a look :)

**Comment by goranbujic** [comment]
How I solved my issue with out-of-center rotation of icon font:
There were several issues I had to resolve:
1. size of icon: it has to be whole px size (e.g. font-size:small; makes my icon 17.5px so center is not absolute center for transformation)
2. defining display: block on icon level makes it center correctly in the middle of the parent div
3. defining exact square size of parent div (in my case button) and fixed padding made it center correctly
4. adding any text-shadow will change the size of the inner icon so it will be out of the center. Trick is to change the hover style to be the same shadow with color same as background in my case

So here is the code:

CSS:
button.close {
    padding: 10px;
    opacity: 0.8;
    text-shadow: 1px 1px 1px #999; 
    /*position: relative;*/
    width: 40px;
    height: 40px;
}
button.close i{
    font-size: 20px;
    max-width: 20px;
    max-height: 20px;
    display: block;
}
button.close:hover {
    text-shadow: 1px 1px 1px #fff; 
}
/* rotation CSS*/
@keyframes anim-rotate {
	0% {
        -moz-transform:rotate(0);
        -webkit-transform:rotate(0);
        transform:rotate(0);
	}
    100% {
        -moz-transform: rotate(360deg);
        -webkit-transform: rotate(360deg);
        transform: rotate(360deg);
    }
}
@keyframes anim-rotate-next {
	0% {
        -moz-transform:rotate(0);
        -webkit-transform:rotate(0);
        transform:rotate(0);
	}
    100% {
        -moz-transform: rotate(360deg);
        -webkit-transform: rotate(360deg);
        transform: rotate(360deg);
    }
}
.rotate{
    animation: anim-rotate-next 1s normal linear;
}
.rotate.down{
    animation: anim-rotate 1s normal linear;
}
HTML:
<button type="button" class="close"><i class="lnr lnr-sync rotate"></i></button>

and finally JQuery for switching rotate class
JQuery:
$("#parentDiv").on('click', 'i.lnr-sync', function () {
            // rotiraj ikonu
            $(this).toggleClass("down");
        });

**Comment by tagliala** [comment]
@goranbujic thanks for sharing this

Could you please make a fiddle showing your approach (you could use this one as a base http://jsfiddle.net/tagliala/4ya23yjL/) and [properly format your post](https://guides.github.com/features/mastering-markdown/)?

Thanks!

**Comment by goranbujic** [comment]
just did it... sorry first one on gitHub :)

Date: Thu, 1 Oct 2015 07:53:45 -0700

**Comment by tomasdev** [comment]
Love you all guys. Thanks for the `transform-origin: 49% 48.5%;` fix on `.fa-refresh.fa-spin` <3

But seriously, would love to get those things symetrically if they're going to be animated!

**Comment by jensgeffken** [comment]
Thanks

**Comment by im4aLL** [comment]
If anyone here for ionicons like me, this will does the trick `transform-origin: 52% 50%;`

**Comment by svbeon** [comment]
confirming the issue on
```
Google Chrome:51.0.2662.0 (Officiell version) canary (64 bitar)
Version	4d223c72fe4e76c7404bffb662f742b947f97107-refs/heads/master@{#378134}
OS	Mac OS X

**Comment by gnail** [comment]
Previously on v4.5.0 I had to add the following CSS style to fix the wobbling issue on spinner:
````
.fa.fa-spinner.fa-pulse {
  transform-origin: 50% 48.5%;
}
````

However on 4.6.1 that css actually made it worse. 4.6.1 seems to have solved the wobbling issue for spinner.

**Comment by My1** [comment]
concerning this one:

http://jsfiddle.net/tagliala/g0ngLhyr/

the question is whether or not this thing is dead center, especially concerning that this thing uses whatever font is available on the system.
it might be intresting to ave an image or an SVG which has a perfect dead-centered circle and let that spin.

**Comment by pilotandy** [comment]
This may or may not help.  I had a wobble in Chrome 51.0.2704.103 m, and realized the css that came with my template had set `letter-spacing: 0.25em;`.  For the fa-spin class, is reset it to `letter-spacing: normal;` and voila, no wobble!

**Comment by bastbnl** [comment]
I'm not sure if we're talking about the same thing but I noticed that `<i class="fa fa-pulse fa-spinner fa-fw"</i>` appears jumpy when used within `small` element. I'm not a css hero, so I just "solved" it by wrapping it within a `span`.

I'm using font awesome 4.6.3 from maxcdn.bootstrap.com with bootstrap 3.3.7 from netdna.bootstrapcnd.com.

**Comment by stenno** [comment]
i can reliably reproduce the wobbling when changing the **font size** of the parent element. In [this fiddle](https://jsfiddle.net/v8kxwkzk/4/) in a chrome browser, the left spinner is wobbling while the right spinner is fixed.

**Comment by Matthew-Bonner** [comment]
The amount of "fixes" people are using worries me, if there are lots of different "fixes" people are posting just how many have actually been tested to work across all browsers.

If the glyph isn't symmetrically positioned such as that the gap between the top and the bottom differ there is going to be a "wobble" as described.

It has already been suggested that remaking the glyphs is a lot of work, so why not just use an image (such as an SVG) which would be better suited for a spinning icon?

**Comment by vizzah** [comment]
Wobbling effect appeared after upgrade from Firefox 48 to Firefox 49 under Ubuntu Xenial 16.04.
Was fixed by addition of "line-spacing: normal" as suggested in this thread.

**Comment by fatalxiao** [comment]
It seems to fix the problem in chrome by adding a style `transform-origin: 48% 51%;`.

Windows 7 / Chrome 53.0.2785.143 m

**Comment by mehrandvd** [comment]
@Matthew-Bonner So, if it is not suitable for this purpose, it should be removed from the documentation!

**Comment by Matthew-Bonner** [comment]
@mehrandvd some of the icons work perfectly fine, hence why there is documentation to cover this, maybe it would be better to update the documentation to make it clear that some icons don't work as well as others and there are known issues.

**Comment by tagliala** [comment]
@Matthew-Bonner this is a good idea and it is partially documented.

> Some browsers on some platforms have issues with animated icons resulting in a jittery wobbling effect. See issue #671 for examples and possible workarounds.

if you have a suggestion to improve this sentence, feel free to submit a PR against the master branch

**Comment by Matthew-Bonner** [comment]
@tagliala I will do, and I think given Font Awesome 5 is being developed, this bug should be put to bed and closed as "Won't fix" as there isn't really anything that can be done.

**Comment by ruwen-m** [comment]
I found out that the wobbling effect depends on the font-size. 
e.g. i changed font size 1.5em to 1.4em and 21px to 20px and voila, no wobbling anymore.

**Comment by Donaldini** [comment]
I solved my issue by replacing the circle-o-notch icon by an svg version of it (with correction of the non-square aspect ratio).

Here's a fiddle for it: [https://jsfiddle.net/vwqzv1mr/5/](https://jsfiddle.net/vwqzv1mr/5/)

**Comment by Taygair** [comment]
I have the same problem (2017) with ALL browsers (IE, FF, Chrome, Safari). Didn't know its such a big problem to keep a spinner centered.

**Comment by visualpharm** [comment]
I'm beginning a campaign, giving away the top requests for FA icons for free, open source. You can incorporate them in a number of ways: generating another icon font, svg set, or (my preference) [compiling](https://iconpharm.com) all FA and third party icons into a single file.

[Different spin icons are here](https://icons8.com/icon/set/spin/all).

Previous ones were [moderator](https://github.com/FortAwesome/Font-Awesome/issues/2553#issuecomment-289439382),  [discord](https://github.com/FortAwesome/Font-Awesome/issues/7574#issuecomment-289875669), [coins](https://github.com/FortAwesome/Font-Awesome/issues/1400#issuecomment-292190852), [Google Play](https://github.com/FortAwesome/Font-Awesome/issues/2272#issuecomment-290106280),  [thin Close](https://github.com/FortAwesome/Font-Awesome/issues/1540#issuecomment-302573662), [walking man](https://github.com/FortAwesome/Font-Awesome/issues/844#issuecomment-302574446), [t-shirt](https://github.com/FortAwesome/Font-Awesome/issues/715), [Xbox](https://icons8.com/icon/12473/xbox), [Nintendo Wii U](https://icons8.com/icon/16297/nintendo-wii-u), [Nintendo](https://icons8.com/icon/16294/nintendo), [PlayStation](https://icons8.com/icon/12461/playstation), [Facebook messenger](https://github.com/FortAwesome/Font-Awesome/issues/6912#issuecomment-302864830) and [Line](https://github.com/FortAwesome/Font-Awesome/issues/2631#issuecomment-292290534) icons.

Please let me know if you want me keep going these icon requests, or compile some upgrade file, or how should we do it?

**Comment by Dhanck7** [comment]
Not shore if this will help but I've managed to make the spinner static by fiddling a bit. I tested both Glyphicons and FontAwesome fonts, just to see witch one looked better. I tried rotating the icon and rotating a container with and icon.  There are 5 key parameters to make this work (font-size, height, width, line height and text indent). 

You can find and example here:
https://jsfiddle.net/Dhanck/syb9qubj/2/

looking forward to version 5!!

**Comment by Taygair** [comment]
@Dhanck7: Thanks for your work! What a CSS orgy is required to make it work! In between I found a simple animation "outside" Font Awesome that works out of the box!

Still can't believe they got over 1 million for Font Awesome 5 but are unable to make a simple animation working properly. *sigh*

**Comment by My1** [comment]
well as you can already see these animations are quite complicated, this doesnt work so easily.

the problem is that we are dealing with characters, aka text, not images which makes this a lot more complicated.

**Comment by Dhanck7** [comment]
And how about this way:
https://jsfiddle.net/Dhanck/9t6y85jx/1/

Just adjusting a 3 parameters and using the default animation behaviour from FontAwesome...

You could also use JavaScript to remove the line gap property I guess.

**Comment by robmadole** [comment]
Alright, everyone.

With version 5 on the horizon I think we've fixed this with the new SVG framework. I just did a battery of tests using the new SVG "cog" and the spin is perfectly centered in the dozens of browsers that I tested.

If you've backed the Kickstarter or pre-ordered and have access to the latest alpha release (available for download from within your account) can you help me test this out?

I'd love to put this issue to bed with Font Awesome 5!

**Comment by Taygair** [comment]
Great to hear! I haven't backed Font Awesome 5 but I'm willing to buy it after release. Do all icons in Font Awesome 5 have the same height and width? It sometimes screwed up my GUI just by different height/width.

**Comment by robmadole** [comment]
@Taygair the icons have the same height but the width varies.

**Comment by Taygair** [comment]
I know... On using the fa-rotate the height turns to width and visa versa. There are some icons that do have - visually - the same height and width (symmetric ones) but they don't. It would be nice if they would have the same height/width (in FA5).

**Comment by afanasy** [comment]
I see that it's probably going to be fixed in the upcoming FA⁵ release, but just in case smb needs a quick fix for the `.fa-circle-o-notch`, here is little dirty css replacement
```css
.loader {
  width: 1em;
  height: 1em;
  border: .1em solid black;
  border-right-color: transparent;
  border-radius: 50%;
  animation: fa-spin 2s infinite linear;
}
```

**Comment by robmadole** [comment]
So far this appears to be working well with the SVG framework. Closing unless someone finds this to be false.

**Comment by tagliala** [comment]
> working well with the SVG framework

If webfonts are still affected and FA5 free will be affected too, I should label this as wontfix and advise to consider a svg solution for spinning icons

**Comment by robmadole** [comment]
@tagliala agreed. I'm afraid that web fonts will always be plagued with issues around this so SVG is the best solution if this is an issue. We'll have FA5 Free out just as soon as we can.

**Comment by My1** [comment]
well SVG framework isnt bad but has one annoying thing: Javascript.

there are many reasons why ppl use things like noscript to protect themselves from stuff js might do and when stuff doesnt work with JS it's ugly imo.

I mean we have things like @font-face and CSS animations and whatnot so that JS is no longer needed for tasks like these and that's a good thing.

**Comment by tagliala** [comment]
Sorry, 😢 

Fonts have a [descender](https://en.wikipedia.org/wiki/Descender) by design, so a perfect spin cannot be guaranteed with the webfont implementation. Labeled as "can't fix"

TL;DR
Can't be fixed because of [descender](https://en.wikipedia.org/wiki/Descender). Please make sure that you are using a font size multiple of 16px (14px for FA4) for your spinner. [**It helps**.](https://jsfiddle.net/tagliala/p4do9La1/2/)
If you are not satisfied and you need a pixel perfect spinner, please consider to use an animated SVG spinner or switch to the SVG implementation

**Comment by PeterShaggyNoble** [comment]
@My1, you don't need JS to animate SVG, you can use CSS.

**Comment by robmadole** [comment]
good call @tagliala ..."can't fix" is more accurate. I've _tried_ multiple times :D

**Comment by My1** [comment]
well but the SVG framework will only load in a JS-enabled browser.

I tried already.

if you have FA Pro, get the demo server, disable js (easy to do in chrome) and access

http://localhost:3344/docs/svg-framework.html

you will see that no icon will load.

you can probably do the svgs alone, but my response was precisely towards the ones who said SVG Framework.

also who has this issue? I tried @tagliala 's link in Firefox chrome and opera and didnt notice much wobble (and I certainly should be able to notice anything weird in the pixels when I go like 5cm close to my 42 inch FHD Screen where you can easily see the pixels)

also foa FA5 would it become 16px then coz the base size changes?

**Comment by TheOneWhoMade** [comment]
how is this a huge problem

**Comment by andreobriennz** [comment]
Using a negative text-indent value fixed the problem for me

**Comment by domenkozar** [comment]
Seems like `filter: blur(0)` does the job and it's very well supported https://caniuse.com/#feat=css-filters

**Comment by workzentre** [comment]
> I see that it's probably going to be fixed in the upcoming FA⁵ release, but just in case smb needs a quick fix for the `.fa-circle-o-notch`, here is little dirty css replacement
> 
> ```css
> .loader {
>   width: 1em;
>   height: 1em;
>   border: .1em solid black;
>   border-right-color: transparent;
>   border-radius: 50%;
>   animation: fa-spin 2s infinite linear;
> }
> ```

Fixed non-decimals sizing fixed it for me too... thanks for sharing.

**Comment by kevbost** [comment]
Hi from the future 👋.

Setting the icon to fixed-width (`fa-fw`) and giving it a `margin: 0 auto;` fixed the wobble for me.

**Comment by camslice** [comment]
@domenkozar 's solution didn't work for me.  This seems to do the trick though:

```
.fa-spin {
    line-height: 0.6em;
}
.fa-spin.fa-lg {
    line-height: 0.7em;
}
.fa-spin.fa-2x {
    line-height: 0.8em;
}
.fa-spin.fa-3x,
.fa-spin.fa-4x,
.fa-spin.fa-5x {
    line-height: 1em;
}
```

As LESS:
```

**Comment by zzztjh** [comment]
> 

> 

> 

> 

> 

> 

> 

> 

> 

> 

> 

> 

> `aaa`

**Comment by noobnoobdc137** [comment]
try
`display: flex;`
to `<i>` tag
![image](https://user-images.githubusercontent.com/55765209/124366711-4ad6be00-dc6f-11eb-93e5-240a64094060.png)
Worked for me

**Comment by camslice** [comment]
@noobnoobdc137 seems to modify the rotation axis and causing the icon to spin over a very wide circle instead of on the spot. Curious about what other CSS you may have applied?

I'm using Font Awesome v4.7.0

**Comment by Alkoptantech** [comment]
Not many problems can be fixed but it looks good at first
https://www.alkoptantech.com

**Comment by Sniper-E** [comment]
Fa-spin had a very slight wobble to it on my site. But fa-pulse had a big wobble to it.

I added a .icon.fa-pulse{padding:2px 0 0 2px} and the wobble was almost completely gone. Now I need to work on fa-spin and slow that slight wobble down.

Try using .icon.fa-spin { padding: (try different numbers) } It will smooth out.

**Comment by tagliala** [comment]
> I added a .icon.fa-pulse{padding:2px 0 0 2px} and the wobble was almost completely gone. Now I need to work on fa-spin and slow that slight wobble down.

It also depends on the browser, so you may experience wobbling on a different browser with the same padding. Please consider an svg based icon if you need a pixel perfect spinner

**Comment by phthatcher-conocophillips** [comment]
2023 - still an issue.  I'm using angular and the '@fortawesome/angular-fontawesome' library, instead of using the fa-spin class, I made a copy of it and added it to my stylesheet, then changed the steps to **1** instead of **8** which fixed the wobble regardless of padding/margin/font-size

I also have transform-origin:center in there somewhere but I don't think its having any effect, it was an earlier attempt to fix the wobble

`<style>

.fa-pulse, .fa-spin-pulse {
  -webkit-animation-name: fa-spin;
  animation-name: fa-spin;
  -webkit-animation-direction: var(--fa-animation-direction, normal);
  animation-direction: var(--fa-animation-direction, normal);
  -webkit-animation-duration: var(--fa-animation-duration, 1s);
  animation-duration: var(--fa-animation-duration, 1s);
  -webkit-animation-iteration-count: var(--fa-animation-iteration-count, infinite);
  animation-iteration-count: var(--fa-animation-iteration-count, infinite);
  -webkit-animation-timing-function: var(--fa-animation-timing, steps(1));
  animation-timing-function: var(--fa-animation-timing, steps(1));
}

</style>`

**Comment by tagliala** [comment]
> 2023 - still an issue

Not an issue, we can't fix this issue, it is by design for webfont implementation

> but the animation-timing steps are set to 1 instead of 8, for me this fixed the wobble regardless if any padding/margin/font-size is set.

This is a fix for some use cases. I would like to specify that fonts have a [descender](https://en.wikipedia.org/wiki/Descender) by design, so a perfect spin cannot be guaranteed with the webfont implementation

It is ok to use a workaround specific for a browser or a use case, but please switch to svg if you need perfect spinners

**Comment by kalamnishkraan** [comment]
Not many problems can be fixed but it looks good at first
[https://www.kalamnishkraan.com](https://www.kalamnishkraan.com)
